### PR TITLE
Add bluetooth firmware and download script

### PR DESCRIPTION
**Pull request by floion**
@agherzan @telphan

### CODE REVIEW

**Comment by agherzan** [comment]
Looks good to me other than having a patch for changes on the components we import directly from Samsung. In this way updating would be easy and we would be able to keep our changes.

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    file://LICENCE.broadcom_bcm43xx \
+    "
+S = "${WORKDIR}"
+
+inherit systemd
+
+SYSTEMD_SERVICE_${PN} = "btfw.service"
+
+RDEPENDS_${PN} = " \
+    bash \
```
Comment (petrosagg) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: Why not use `/bin/sh`? No need to install bash in our images

Comment (floion) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: Bash is already installed in our images

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    file://hciconf.sh \
+    "
+S = "${WORKDIR}"
+
+inherit systemd
+
+SYSTEMD_SERVICE_${PN} = "brcm-bt-firmware.service"
+
+RDEPENDS_${PN} = " \
+    bash \
```
Comment (agherzan) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: You don't have any specific bash dependencies. You can use simple sh.

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            install -d ${D}${systemd_unitdir}/system
+            install -c -m 0644 ${WORKDIR}/brcm-bt-firmware.service ${D}${systemd_unitdir}/system
+            sed -i -e 's,@BASE_BINDIR@,${base_bindir},g' \
+                -e 's,@BASE_SBINDIR@,${base_sbindir},g' \
+                -e 's,@SBINDIR@,${sbindir},g' \
+                -e 's,@BINDIR@,${bindir},g' \
+                -e 's,@SYS_CONFDIR@,${sysconfdir},g' \
+                ${D}${systemd_unitdir}/system/brcm-bt-firmware.service
+        fi
+
```
Comment (agherzan) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: NIT: newline can be removed

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+PACKAGE_ARCH = "${MACHINE_ARCH}"
```
Comment (agherzan) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: Is this really a binary of the machine architecture? I don't think this is really arch dependent. Most probably it's arch all and compatible only for artikX

File: recipes-bsp/bcm4354-bluetooth/files/10-local.rules [hunk 5]
```diff
@@ -0,0 +1 @@
+ACTION=="add", KERNEL=="hci0", RUN+="/etc/bluetooth/hciconf.sh"
```
Comment (agherzan) [inline recipes-bsp/bcm4354-bluetooth/files/10-local.rules]: EOF new line missed.

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+SYSTEMD_SERVICE_${PN} = "brcm-bt-firmware.service"
+
+RDEPENDS_${PN} = " \
+    bash \
+    bluez5 \
+    "
+
+do_install() {
+        install -d  ${D}/etc/bluetooth/
+        install -m 0755 ${WORKDIR}/BCM4354_003.001.012.0353.0745_Samsung_Artik_ORC.hcd ${WORKDIR}/brcm_patchram_plus ${WORKDIR}/fwdown.sh ${WORKDIR}/hciconf.sh ${D}/etc/bluetooth/
```
Comment (agherzan) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: Installing executable and scripts in etc looks strange.

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb (hunk 2 above)
Comment (floion) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: /bin/sh is a symlink to /bin/bash in our rootfs

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb (hunk 4 above)
Comment (floion) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: It's on the same chipset as the wifi, same bcm4354. We agreed that the wifi changes was machine specific. In which case this also should be

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb (hunk 3 above)
Comment (floion) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: Make sure you review the latest files. This is not the latest file by the way. And I would prefer to keep the newline

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb (hunk 6 above)
Comment (floion) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: I followed the Samsung Fedora way

File: recipes-bsp/bcm4354-bluetooth/files/10-local.rules (hunk 5 above)
Comment (floion) [inline recipes-bsp/bcm4354-bluetooth/files/10-local.rules]: Will add newline

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb (hunk 2 above)
Comment (floion) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: Will remove the bash runtime dep

File: recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb (hunk 3 above)
Comment (agherzan) [inline recipes-bsp/bcm4354-bluetooth/bcm4354-bt.bb]: What do you mean? I wanted to say that the function block has an extra newline before closing the block.

**Pull request merged**